customizable host key policy

anarcat opened the pull request:
The current Fabric code does not check or save SSH host keys at all. This PR fixes the problem by ensuring we properly call the Paramiko SSH keys loader when we create our connection object.

We also load the right UserKnownHosts file based on the ssh_config, which paramiko does not do out of the box.

We might want to push the latter into paramiko, but first it seems to me the default Fabric policy of "everything goes" should b fixed.

Note that this also allows callers to override the AutoAddPolicy that is currently in use.

This should fix #2071

This is a rebase of #2072 on top of master (instead of 2.0, see #1929 for that) and with #1960 merged (instead of #2073) to fix the docs.

anarcat committed:
make host key policy customizable

This is essential in case we want Fabric library consumers and users
to override the Paramiko host key policy. One, for example, might want
to override the default to prompt users when a new key needs to be
added (which is silently done by default right now).

anarcat committed:
default host key policy to paramiko's

This removes the Fabric-specific host key policy which is to
automatically accept new keys. I don't see why Fabric would do
anything different than the default here at this stage, especially
since we don't seem to be *loading* any keys, let alone saving them.

This is a stopgap measure to fix #2071

anarcat committed:
load the right host key according to ssh_config rules (#2071)

This has the following consequences:

 1. the AutoAddPolicy starts to work correctly, as the
    `load_host_keys` call signals that it should save new fingerprints
    to the known_hosts file

 2. it will also correctly *fail* if the fingerprint on a host
    changes, which was *not* the case before, as the known_hosts file
    was never loaded

 3. it will correctly load the right `UserKnownHostsFile` according to
    the rules defined in the ssh configuration

Regarding the last point, that is essential in my setup because I have
a config block like this:

Host *.example.com
     UserKnownHostsFile ~/.ssh/known_hosts.example.com

... which allows me to "group" known_hosts file and synchronize them
from remote servers.

anarcat committed:
allow multiple known_hosts keys to be provided

This covers the following use case:

    Host *.example.org
         UserKnownHostsFile ~/.ssh/known_hosts ~/.ssh/known_hosts.example.org

... which is also valid.

anarcat committed:
do not crash on missing known hosts keys file

anarcat committed:
fix unit tests, now that the key policy is changeable

anarcat committed:
wrap SSH client setup in a function

This makes it easier to monkeypatch to earlier versions. With this
patch, I can copy-paste the setup_ssh_client() function into my code
and monkeypatch older fabric versions with:

    # hack to fix Fabric key policy:
    # https://github.com/fabric/fabric/issues/2071
    def safe_open(self):
        SaferConnection.setup_ssh_client(self)
        Connection.open_orig(self)

    class SaferConnection(Connection):
        # this function is a copy-paste from https://github.com/fabric/fabric/pull/2072
        def setup_ssh_client(self):
            # [...]

    Connection.open_orig = Connection.open
    Connection.open = safe_open

This is otherwise a noop.

d1b commented:
cc @bitprophet . 

As an aside I think it is well overdue for fabric to offer a sensible & secure default for users and not blindly auto-add ssh keys. Of course changing the default behaviour would require a new major (if following semver) version bump.

anarcat committed:
default to RejectPolicy

The argument here is that it's a really bad idea to silently accept
new keys. It's not how SSH works, and it's not how Fabric *should*
work. If we'd work like SSH, we'd *prompt* the user, but the
WarningPolicy doesn't do that, it just warns.

At the very least those policies should noisily warn the user when the
host keys get changed, but that's something Paramiko must do.

tacerus committed:
Assert new default host key policy

Since the default policy was changed, reference the new value to
repair the test assertion.

Signed-off-by: Georg Pfuetzenreuter <georg.pfuetzenreuter@suse.com>

tacerus commented:
Thank you for the patch, upon trying to apply this on top of our distribution package, I noticed the test suite to fail due to still missing a minor adjustment which I added here: https://github.com/tacerus/fabric/commit/9c915064babc30daed6ea22431ab02d3626d5d3c (I suppose it could alternatively be read from `default_host_key_policy` now).

tacerus commented:
It seems some Python 3.6 tests fail here in the upstream pipeline now as a result, sorry about that, I did not test with 3.6. :-(
